### PR TITLE
[bcq-tools] Introduce preserving BCQ information tool

### PR DESCRIPTION
**Pull request by llFreetimell**
Parent Issue : #3471 

This commit will introduce `preserve_bcq_info` tool, which is for preserving BCQ information.

ONE-DCO-1.0-Signed-off-by: Seok NamKoong <seok9311@naver.com>

### CODE REVIEW

File: compiler/bcq-tools/preserve_bcq_info [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+    original_graph_model_def = original_graph_model.as_graph_def()
+
+    new_graph = tf.compat.v1.GraphDef()
+    substitution_dict = {}
+
+    DT_INT32 = None  # Just for copying DT_INT32 attribute value
+
+    for node in original_graph_model_def.node:
+        if node.op == "Const":
+
```
Comment (mhs4670go) [inline compiler/bcq-tools/preserve_bcq_info:75]: maybe redundant space?

Comment (llFreetimell) [inline compiler/bcq-tools/preserve_bcq_info:75]:
```suggestion
```

File: compiler/bcq-tools/README.md [hunk 2]
```diff
@@ -0,0 +1,44 @@
+# BCQ Tools
+
+This directory includes some tools related with BCQ.
+
+## preserve_bcq_info
+
+### Purpose
+
+`preserve_bcq_info` is for preserving constant nodes which are including BCQ information.
```
Comment (seanshpark) [inline compiler/bcq-tools/README.md:9]: (optional) `which are including ` -> `which includes`

File: compiler/bcq-tools/README.md [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+When `.pb` file is converted to `.tflite` file by TFlite converter, constant nodes whose values are exactly same are removed and then linked to only one representative node.
+It causes information missing problem because we don't know which constant nodes should be linked to even we still want to apply BCQ.
```
Comment (seanshpark) [inline compiler/bcq-tools/README.md:11]: maybe...
`This causes lost of information problem as we don't know which constant node should be linked to the node that we want to apply BCQ`

File: compiler/bcq-tools/README.md (hunk 2 above)
Comment (llFreetimell) [inline compiler/bcq-tools/README.md:9]:
```suggestion
`preserve_bcq_info` is for preserving constant nodes which include BCQ information.
```

File: compiler/bcq-tools/README.md [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+
+This directory includes some tools related with BCQ.
+
+## preserve_bcq_info
+
+### Purpose
+
+`preserve_bcq_info` is for preserving constant nodes which include BCQ information.
+When `.pb` file is converted to `.tflite` file by TFlite converter, constant nodes whose values are exactly same are removed and then linked to only one representative node.
+It causes information missing problem because we don't know which constant nodes should be linked to even we still want to apply BCQ.
```
Comment (llFreetimell) [inline compiler/bcq-tools/README.md:11]:
```suggestion
This makes us impossible to know what constant node should be linked to a node which we want to apply BCQ.
```

File: compiler/bcq-tools/README.md [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+This directory includes some tools related with BCQ.
+
+## preserve_bcq_info
+
+### Purpose
+
+`preserve_bcq_info` is for preserving constant nodes which include BCQ information.
+When `.pb` file is converted to `.tflite` file by TFlite converter, constant nodes whose values are exactly same are removed and then linked to only one representative node.
+This makes us impossible to know what constant node should be linked to a node which we want to apply BCQ.
+Therefore, we should preserve these BCQ information.
```
Comment (llFreetimell) [inline compiler/bcq-tools/README.md:12]:
```suggestion
One of the solutions is making all the same constant nodes different by inserting unique values and ignore the newly generated unique values when BCQ fusing is applied.
`preserve_bcq_info` will generate and insert unique dummy values to the constant nodes whose values are same not to be removed by Tensorflow Lite converter.
As a result, BCQ information will be preserved.
```

File: compiler/bcq-tools/preserve_bcq_info [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+            # Because bcqinfo_do_w_x is BOOL type, we cannot add dummy value at the end.
+            # Therefore we should convert the type to INT32 type.
+            if "/bcqinfo_do_w_x" in node.name:
+                original_tensor = tf.make_ndarray(node.attr["value"].tensor)
+                substitution_dict[node.name] = tf.make_tensor_proto(
+                    [int(original_tensor[0]), unique_value.gen()], tf.int32)
+
+            if ("/bcqinfo_number_of_clusters" in node.name
+                    or "/bcqinfo_size_of_clusters" in node.name
+                    or "/bcqinfo_qbits_of_clusters" in node.name):
```
Comment (mhs4670go) [inline compiler/bcq-tools/preserve_bcq_info:84]: (Optional)
```suggestion
            bcqinfo_name = [ "/bcqinfo_number_of_clusters", "/bcqinfo_size_of_clusters", "/bcqinfo_qbits_of_clusters"]
            if any(name in node.name for name in bcqinfo_name):
```

**Comment by seanshpark** [comment]
@llFreetimell , have you tested the new code?

**Comment by llFreetimell** [comment]
@seanshpark Yes, I have used this code to generate BCQ applied model file.

**Comment by mhs4670go** [comment]
Just curiosity, I've tested with exactly same value constants.
```python
import tensorflow as tf

lhs_ = tf.compat.v1.placeholder(dtype=tf.float32, shape=(4, 4), name="Hole")
rhs_ = tf.compat.v1.placeholder(dtype=tf.float32, shape=(4, 4), name="Hole")
scale = tf.compat.v1.constant([1., 2., 3.], name="const1")
scale2 = tf.compat.v1.constant([1., 2., 3.], name="const2")
op_ = tf.compat.v1.add(lhs_, rhs_)

$ ~/ONE/build/compiler/tf2tfliteV2/tf2tfliteV2.py --v1 -i add.pbtxt -o add.tflite -I Hole,Hole_1 -O Add,const1,const2
```
But 
image

const1, const2 still alive. Is there something missing?

**Comment by seanshpark** [comment]
> const1, const2 still alive. Is there something missing?

You know best :)

**Pull request merged**